File: Codeforces/Contests/Educational 167/D.py
n, m = list(map(int, input().split()))
A = list(map(int, input().split())) # cost of crafting
B = list(map(int, input().split())) # return of smelting
C = list(map(int, input().split())) # no. of ingots of each type you start with

D = [0]*1000000 # D[cost of one craft/smelt cycle] = min required threshold to craft
for i in range(n):
    t = A[i] - B[i]
    if D[t] != 0:
        if D[t] > A[i]:
            D[t] = A[i]
    else:
        D[t] = A[i]
Ds = [(c, t) for c, t in enumerate(D) if t > 0]
i = 0
while (i < len(Ds) - 1):
    (k1, v1), (k2, v2) = Ds[i], Ds[i+1]
    if v1 < v2:
        Ds.pop(i+1)
    else:
        i += 1
D = None

memo = [0]*1000000

res = 0
for j in range(m):
    c = C[j]
    
    # First weapon:
    k, v = Ds[0]
    if (c < v):
        True
    else:
        r = (c - v) // k + 1
        res += 2 * r
        c -= (r * k)
    
    # Second weapon onward: memoise
    c1, t1 = c, 0
    if (memo[c] != 0):
        t1 += memo[c]
    else:
        for k, v in Ds[1:]:
            # print(f"{k}, {v}, {j}, {c}")
            if (c < v):
                if (c < k):
                    break
                continue
            else:
                r = (c - v) // k + 1
                t1 += 2 * r
                c -= (r * k)
        memo[c1] = t1
    res += t1
print(res)

File: Codeforces/Contests/Educational 167/D.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

int main() {
    int n; int m; scanf("%i %i", &n, &m);
    vector<ll> A(n, 0);
    vector<ll> B(n, 0);
    vector<ll> C(m, 0);
    for (int i=0; i<n; i++) { scanf("%lli", &A[i]); }
    for (int i=0; i<n; i++) { scanf("%lli", &B[i]); }
    for (int i=0; i<m; i++) { scanf("%lli", &C[i]); }
    map<int, int> D;
    for (int i=0; i<n; i++) {
        ll t = A[i] - B[i];
        if (D.count(t) > 0) {
            if (D[t] > A[i]) D[t] = A[i];
        } else {
            D[t] = A[i];
        }
    }
    auto it = D.begin();
    while (it != prev(D.end())) {
        if (it->second < (next(it))->second) D.erase(next(it));
        else ++it;
    }

    vector<int> memo(1000000, 0);
    ll res = 0;
    for (int j=0; j<m; j++) {
        ll ingots = C[j], part_res = 0ll;
        auto it = D.begin();
        vector<pair<ll, ll>> T;
        while (it != D.end()) {
            ll cost = it->first, threshold = it->second;
            if (ingots < threshold) {
                if (ingots < cost) break;
                it++;
            } else {
                ll r = (ingots - threshold) / cost + 1;
                T.emplace_back(pair<ll, ll>(2 * r, ingots));
                ingots -= (r * cost);
                if (memo[ingots] != 0) {
                    part_res += memo[ingots];
                    break;
                }
                it++;
            }
        }
        ll l = T.size();
        if (l > 0) {
            while (l > 1) {
                l--;
                auto p = T[l];
                part_res += p.first;
                memo[p.second] = part_res;
            }
            res += (part_res + T[0].first);
        }
    }
    printf("%lli", res);

}